feat: add unleash web provider

jarebudev opened the pull request:
## This PR

- adds an unleash provider

### Related Issues

Fixes #1069 

### Notes

### Follow-up Tasks
- Missing unit tests to verify that the context is updated - will add these

### How to test

- There are Jest unit tests

lukas-reining commented:
Hey @jarebudev I will have a look at this in the next days. Sorry for the delay :)

jarebudev commented:
> Hey @jarebudev I will have a look at this in the next days. Sorry for the delay :)

No worries :)

lukas-reining commented on libs/providers/unleash-web/src/lib/unleash-web-provider.ts:
```diff
@@ -0,0 +1,164 @@
+    }
+    return {
+      variant: variant,
+      value: value as T,
```
Please add a check if the type is convertible to the type `T`, you can do it using the default value. 

E.g. for numbers, add something like: 
```ts
if (typeof defaultValue === Number && isNaN(parseFloat(value))) {
  throw new TypeMismatchError()
} else if (typeof defaultValue ===
.....
```

And then instead of doing the conversion in the specific methods you can do them here to ensure type safety at a central place.

jarebudev replied:
yes makes sense, I will look into this tomorrow.

lukas-reining reviewed (commented):
Sorry for the delay again, I had quite busy days. 
This looks really good! :) 
I added a comment regarding the type checking and conversion, please have a look if you agree. 
Tomorrow I will have  another look but this already looks pretty good.

Edit: I will have a look at the failing tests tomorrow too.

jarebudev commented on libs/providers/unleash-web/src/lib/unleash-web-provider.ts:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    if (evaluatedVariant.name === 'disabled') {
+      value = defaultValue as T;
```
almost certainly not but I will double check!

jarebudev commented:
> Sorry for the delay again, I had quite busy days. This looks really good! :) I added a comment regarding the type checking and conversion, please have a look if you agree. Tomorrow I will have another look but this already looks pretty good.
> 
> Edit: I will have a look at the failing tests tomorrow too.

Thanks!
I've added a change to remove the extra cast.

I noticed that the CI is failing and I was unable to replicate the build failures locally.  Have I overlooked something?

jarebudev commented on .release-please-manifest.json:
```diff
@@ -18,5 +18,6 @@
   "libs/providers/config-cat-web": "0.1.3",
-  "libs/shared/config-cat-core": "0.1.0"
+  "libs/shared/config-cat-core": "0.1.0",
+  "libs/providers/unleash-web": "0.1.0"
```
good spot :)

jarebudev commented on libs/providers/unleash-web/package.json:
```diff
@@ -0,0 +1,17 @@
+  "version": "0.0.1",
+  "dependencies": {
+    "tslib": "^2.3.0",
+    "unleash-proxy-client": "^3.6.1"
```
Thanks - I've applied the changes that you suggested and this has fixed the CI build!

lukas-reining reviewed (approved):
Thank you @jarebudev!
This looks good to me now!

beeme1mr commented on libs/providers/unleash-web/README.md:
````diff
@@ -0,0 +1,82 @@
+After the provider gets initialized, you can start evaluations of feature flags like so:
+
+```ts
+// Note - this can also be set within the contructor
````
It's best practice to include this in the constructor to avoid unnecessary round trips.

beeme1mr commented on libs/providers/unleash-web/src/lib/unleash-web-provider.ts:
```diff
@@ -0,0 +1,187 @@
+
+  async initialize(): Promise<void> {
+    await this.initializeClient();
+    this._logger?.info('UnleashWebProvider initialized');
```
It may be better to use debug into of info since this is a library but it's up to you.

jarebudev replied:
no worries i'll change it so it logs at debug level

beeme1mr reviewed (approved):
Looks good so far. My only concern is the log level but it's not a blocker for me. Thanks!

toddbaert reviewed (approved):
Approved - I just reverted some submodule changes in flagd, which I'm sure weren't intentional.
